Exceptions raised in a Directive cause parallel builds to hang

**Describe the bug**

While it seems that raising exceptions in directives is at least frowned upon, the behaviour of when this happens during a parallel build seems to not be what I would expect.

If a (custom) directive raises an exception (regular Python exception, or sphinx ExtensionError) the parallel build will hang indefinitely after printing the error to screen.

**To Reproduce**

Modify an extension to raise an error inside the `run()` method of a Directive class, then build the project with `-j auto`.

**Expected behaviour**
The build process exits with a failing code after the error is displayed.

**Environment info**
- OS: Linux
- Python version: 3.9
- Sphinx version: 3.3.2
- Sphinx extensions:  https://github.com/OpenAstronomy/sphinx-changelog but also reproduceable with other Directives

Is it really? I suppose Sphinx handles the error on the parallel build. As a trial, I added the following directive into conf.py:
```
from sphinx.util.docutils import SphinxDirective

class MyDirective(SphinxDirective):
    def run(self):
        raise

def setup(app):
    app.add_directive('my', MyDirective)
```

And sphinx-build exits successfully with displaying an error.
```
$ make clean html SPHINXOPTS=-jauto
Removing everything under '_build'...
Running Sphinx v3.5.1
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 14 source files that are out of date
updating environment: [new config] 14 added, 0 changed, 0 removed
reading sources... [100%] subdir/foo .. subdir/qux
waiting for workers...

Sphinx parallel build error:
RuntimeError: No active exception to reraise
make: *** [Makefile:20: html] Error 2
```

Is there any example project?

--- a/CHANGES
+++ b/CHANGES
@@ -16,6 +16,8 @@ Features added
 Bugs fixed
 ----------
 
+* #8952: Exceptions raised in a Directive cause parallel builds to hang
+
 Testing
 --------
 

--- a/sphinx/util/parallel.py
+++ b/sphinx/util/parallel.py
@@ -103,8 +103,21 @@ class ParallelTasks:
         self._join_one()
 
     def join(self) -> None:
-        while self._pworking:
-            self._join_one()
+        try:
+            while self._pworking:
+                self._join_one()
+        except Exception:
+            # shutdown other child processes on failure
+            self.terminate()
+            raise
+
+    def terminate(self) -> None:
+        for tid in list(self._precvs):
+            self._procs[tid].terminate()
+            self._result_funcs.pop(tid)
+            self._procs.pop(tid)
+            self._precvs.pop(tid)
+            self._pworking -= 1
 
     def _join_one(self) -> None:
         for tid, pipe in self._precvs.items():

